File: services/meeting/src/main/java/org/transhub/exception/ErrorCode.java
package org.transhub.exception;

import org.springframework.http.HttpStatus;

public enum ErrorCode {
    UNCATEGORIZED_EXCEPTION(9999, "Uncategorized error", HttpStatus.INTERNAL_SERVER_ERROR),
    INVALID_KEY(1001, "Uncategorized error", HttpStatus.BAD_REQUEST),
    UNAUTHENTICATED(1006, "Unauthenticated", HttpStatus.UNAUTHORIZED),
    UNAUTHORIZED(1007, "You do not have permission", HttpStatus.FORBIDDEN),

    // Meeting Service error codes
    MEETING_NOT_FOUND(5001, "Meeting not found", HttpStatus.NOT_FOUND),
    AUDIO_FILE_NOT_FOUND(5002, "Audio file not found", HttpStatus.NOT_FOUND),
    MEMBER_NOT_FOUND(5003, "Member not found in meeting", HttpStatus.NOT_FOUND),
    MEMBER_ALREADY_EXISTS(5004, "User is already a member of this meeting", HttpStatus.BAD_REQUEST),
    INVALID_ACTION(5005, "Invalid action or permission", HttpStatus.BAD_REQUEST),
    USER_NOT_FOUND(5006, "User not found", HttpStatus.NOT_FOUND),
    AUDIO_FILE_ALREADY_LINKED(5007, "Audio file is already linked to another meeting", HttpStatus.BAD_REQUEST),
    ;

    ErrorCode(int errorCode, String message, HttpStatus httpStatus) {
        this.errorCode = errorCode;
        this.message = message;
        this.httpStatus = httpStatus;
    }

    private final HttpStatus httpStatus;
    private final int errorCode;
    private final String message;

    public int getErrorCode() {
        return errorCode;
    }

    public String getMessage() {
        return message;
    }

    public HttpStatus getHttpStatus() {
        return httpStatus;
    }
}

File: services_ms/libs/common/src/exceptions/error-code.ts
import { HttpStatus } from '@nestjs/common';

export interface ErrorCodeInfo {
  errorCode: number;
  message: string;
  httpStatus: HttpStatus;
}

export const ErrorCodes = {
  UNCATEGORIZED_EXCEPTION: { errorCode: 9999, message: 'Uncategorized error', httpStatus: HttpStatus.INTERNAL_SERVER_ERROR },
  INVALID_KEY: { errorCode: 1001, message: 'Validation failed', httpStatus: HttpStatus.BAD_REQUEST },
  USER_NOT_EXISTED: { errorCode: 1005, message: 'User not existed', httpStatus: HttpStatus.NOT_FOUND },
  UNAUTHENTICATED: { errorCode: 1006, message: 'Unauthenticated', httpStatus: HttpStatus.UNAUTHORIZED },
  UNAUTHORIZED: { errorCode: 1007, message: 'You do not have permission', httpStatus: HttpStatus.FORBIDDEN },
  EMAIL_EXISTED: { errorCode: 1009, message: 'Email existed, please choose another one', httpStatus: HttpStatus.BAD_REQUEST },
  USER_EXISTED: { errorCode: 1010, message: 'Username existed, please choose another one', httpStatus: HttpStatus.BAD_REQUEST },
  USER_NOT_FOUND: { errorCode: 1005, message: 'User not found', httpStatus: HttpStatus.NOT_FOUND },
  ROLE_NOT_FOUND: { errorCode: 1011, message: 'Role not found', httpStatus: HttpStatus.NOT_FOUND },
  ROLE_NAME_EXISTED: { errorCode: 1012, message: 'Role name already exists', httpStatus: HttpStatus.BAD_REQUEST },
  
  // File Service
  FILE_NOT_FOUND: { errorCode: 4001, message: 'File not found', httpStatus: HttpStatus.NOT_FOUND },
  FILE_EMPTY: { errorCode: 4002, message: 'Uploaded file is empty', httpStatus: HttpStatus.BAD_REQUEST },
  INVALID_RANGE: { errorCode: 4008, message: 'Requested range is not satisfiable', httpStatus: HttpStatus.REQUESTED_RANGE_NOT_SATISFIABLE },
  AUDIO_FILE_NOT_FOUND: { errorCode: 4003, message: 'Audio file not found', httpStatus: HttpStatus.NOT_FOUND },
  FILE_UPLOAD_FAILED: { errorCode: 4004, message: 'File upload failed', httpStatus: HttpStatus.INTERNAL_SERVER_ERROR },
  FILE_MERGE_FAILED: { errorCode: 4005, message: 'File merge failed', httpStatus: HttpStatus.INTERNAL_SERVER_ERROR },

  // Meeting Service
  MEETING_NOT_FOUND: { errorCode: 5001, message: 'Meeting not found', httpStatus: HttpStatus.NOT_FOUND },
  MEMBER_NOT_FOUND: { errorCode: 5003, message: 'Member not found in meeting', httpStatus: HttpStatus.NOT_FOUND },
  MEMBER_ALREADY_EXISTS: { errorCode: 5004, message: 'User is already a member of this meeting', httpStatus: HttpStatus.BAD_REQUEST },
  INVALID_ACTION: { errorCode: 5005, message: 'Invalid action or permission', httpStatus: HttpStatus.BAD_REQUEST },
  AUDIO_FILE_ALREADY_LINKED: { errorCode: 5007, message: 'Audio file is already linked to another meeting', httpStatus: HttpStatus.BAD_REQUEST },

  // Transcript Service
  TRANSCRIPT_NOT_FOUND: { errorCode: 6001, message: 'Transcript not found', httpStatus: HttpStatus.NOT_FOUND },
  TRANSCRIPT_ALREADY_EXISTS: { errorCode: 6002, message: 'Transcript already exists for this audio file', httpStatus: HttpStatus.BAD_REQUEST },
};

export class AppException extends Error {
  constructor(
    public readonly errorCodeInfo: ErrorCodeInfo,
    message?: string,
  ) {
    super(message || errorCodeInfo.message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
